prompb/write/v2/types.proto

```diff
@@ -18,47 +18,84 @@ option go_package = "writev2";
+  // as indices to the WriteRequest.symbols array. This list's len is always
+  // a multiple of 2.
+  //
+  // Note that there might be multiple TimeSeries objects in the same
+  // WriteRequests with the same labels e.g. for different exemplars, metadata
+  // or created timestamp.
+  repeated uint32 labels_refs = 1;
+  // samples contain zero or more samples for a given timeseries. For a typical
+  // clients, in healthy cases, there will be only one sample, for ~real
+  // time metric streaming.
+  //
+  // Samples are sorted by time (older first).
+  // TODO(bwplotka): Can/should samples and histograms be in enum? Do we allow mix of those in one TimeSeries?
   repeated Sample samples = 2 [(gogoproto.nullable) = false];
+  // exemplars represents optional set of exemplars attached to this series' samples.
   repeated Exemplar exemplars = 3 [(gogoproto.nullable) = false];
-  // Same as current version
+  // histograms represents zero or more histogram samples for a this series.
+  // TODO(bwplotka): ditto
+  // Histogram samples are sorted by time (older first).
   repeated Histogram histograms = 4 [(gogoproto.nullable) = false];
-
+  // metadata represents metadata associated with the given series' samples.
   Metadata metadata = 5 [(gogoproto.nullable) = false];
-
-  // Optional created timestamp for the metric in ms format,
-  // if the first sample in samples does not contain 0 value.
-  // See model/timestamp/timestamp.go for conversion from time.Time
-  // to Prometheus timestamp.
+  // created_timestamp represents an optional created timestamp associated with
+  // this series' samples in ms format, typically for counter or histogram type
+  // metrics. NOTE: Servers might negotiate "required-created-timestamp" feature
+  // that requires this field to be set for counters and histograms.
+  // For Go, see github.com/prometheus/prometheus/model/timestamp/timestamp.go
+  // for conversion from/to time.Time to Prometheus timestamp.
   int64 created_timestamp = 6;
 }
 
+// Exemplar is an additional information attached to some series' samples.
+// It is typically used to attach an example trace or request ID associated with
+// the metric changes.
 message Exemplar {
-  // TODO: same as TimeSeries.labels_refs
+  // labels_refs is a sorted list of label name-value pair references, encoded
+  // as indices to the WriteRequest.symbols array. This list's len is always
+  // a multiple of 2.
   repeated uint32 labels_refs = 1;
+  // value represents an exact example value. This can be useful when the exemplar
+  // is attached to a histogram, which only gives an estimated value through buckets.
   double value = 2;
-  // timestamp is in ms.
+  // timestamp represents an optional timestamp of the example in ms.
+  // For Go, see github.com/prometheus/prometheus/model/timestamp/timestamp.go
+  // for conversion from/to time.Time to Prometheus timestamp.
   int64 timestamp = 3;
```

Comment:
Same for optionality with these two fields.

Reply:
Mmm giving this a second thought - AFAIU it makes sense to make `timestamp` and `value` optionals for exemplars, but I wonder how much impact it would have on existing backends, given that these exist in RW 1.0 and are not optional. Maybe it complicated the migration?

Reply:
according to the spec [here](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#exemplars) exemplars must have a labelset and a value, only the timestamp is optional